Core/clim-basic/drawing/graphics.lisp

```diff
@@ -600,74 +600,77 @@
+                         line-style line-thickness
+                         line-unit line-dashes line-cap-shape
+                         (to-head t) from-head (head-length 10) (head-width 5) angle)
   (declare (ignore ink clipping-region transformation
                    line-style line-thickness
                    line-unit line-dashes line-cap-shape))
   (with-medium-options (sheet args)
     (with-translation (sheet x2 y2)
-      (with-rotation (sheet (or angle (atan* (- x1 x2)
-                                             (- y1 y2))))
-        (let* ((end 0.0)
-               (start (sqrt (+ (expt (- x2 x1) 2)
-                               (expt (- y2 y1) 2))))
-               (p end)
-               (q start)
-               (line-style (medium-line-style sheet))
-               ;; FIXME: I believe this thickness is in "line-style-units",
-               ;; which are only coincidentally the same as pixel coorindates
-               ;; on screen backends, using :normal units. There is no function
-               ;; documented for converting the units to stream coordinates.
-               (thickness (multiple-value-bind (dx dy)
-                              (transform-distance
-                               (invert-transformation
-                                (medium-transformation sheet))
-                               (line-style-thickness line-style)
-                               0)
-                            (sqrt (+ (* dx dx) (* dy dy)))))
-               (width/2 (/ head-width 2))
-               (a (atan (/ width/2 head-length)))
-               (offset (if (and head-length (not (zerop head-length)))
-                           (/ thickness (* 2 (sin a )))
-                           0.0))
-               (tip-to-peak (+ head-length
-                               offset
-                               (- (* thickness 0.5 (sin a)))))) ;; okay, a guess..
-          (when to-head   (incf p offset))
-          (when from-head (decf q offset))
-          (if (and to-head
-                   from-head
-                   (< (abs (- start end)) (* 2 tip-to-peak)))
-              (let ((width (* 0.5 (+ head-width thickness)
-                              (/ (abs (- start end))
-                                 (* 2 tip-to-peak)) )))
-                (draw-polygon* sheet
-                               (list end 0
-                                     (/ start 2) width
-                                     start 0
-                                     (/ start 2) (- width))
-                               :filled t
-                               :line-thickness 0))
-              (progn
-                (when to-head
+      (let ((delta-x (- x1 x2))
+            (delta-y (- y1 y2)))
+        (with-rotation (sheet (or angle (if (and (zerop delta-x) (zerop delta-y))
+                                            0.0
+                                            (atan* delta-x delta-y))))
```

Comment:
maybe something like a proposition below? notice, that there is no need to compute delta when angle is present.

```
  (with-medium-options (sheet args)
    (with-translation (sheet x2 y2)
      (unless angle
        (let ((dx (- x1 x2))
              (dy (- y1 y2)))
          (if (and (zerop dx)
                   (zerop dy))
              (setf angle 0.0)
              (setf angle (atan* dx dy)))))
      (with-rotation (sheet angle)
        (let* ((end 0.0)
               (start (sqrt (+ (expt (- x2 x1) 2)
                               (expt (- y2 y1) 2))))
               (p end)
               (q start)
```

Reply:
done that